http-server/src/main/java/io/micronaut/http/server/cors/CorsFilter.java

```diff
@@ -430,39 +451,37 @@ private void decorateResponseWithHeaders(@NonNull HttpRequest<?> request,
 
     @NonNull
     private MutableHttpResponse<?> handlePreflightRequest(@NonNull HttpRequest<?> request,
-                                                                     @NonNull CorsOriginConfiguration corsOriginConfiguration) {
-        Optional<HttpStatus> statusOptional = validatePreflightRequest(request, corsOriginConfiguration);
-        if (statusOptional.isPresent()) {
-            HttpStatus status = statusOptional.get();
-            if (status.getCode() >= 400) {
-                return HttpResponse.status(status);
-            }
-            MutableHttpResponse<?> resp = HttpResponse.status(status);
-            decorateResponseWithHeadersForPreflightRequest(request, resp, corsOriginConfiguration);
-            decorateResponseWithHeaders(request, resp, corsOriginConfiguration);
-            return resp;
+                                                          @NonNull CorsOriginConfiguration corsOriginConfiguration) {
+        HttpStatus status = validatePreflightRequest(request, corsOriginConfiguration);
+        if (status.getCode() >= 400) {
+            return HttpResponse.status(status);
         }
-        return null;
+        MutableHttpResponse<?> resp = HttpResponse.status(status);
+        decorateResponseWithHeadersForPreflightRequest(request, resp, corsOriginConfiguration);
+        decorateResponseWithHeaders(request, resp, corsOriginConfiguration);
+        return resp;
     }
 
     @NonNull
-    private Optional<HttpStatus> validatePreflightRequest(@NonNull HttpRequest<?> request,
-                                                          @NonNull CorsOriginConfiguration config) {
+    private HttpStatus validatePreflightRequest(@NonNull HttpRequest<?> request,
+                                                @NonNull CorsOriginConfiguration config) {
         Optional<HttpMethod> methodToMatchOptional = validateMethodToMatch(request, config);
         if (methodToMatchOptional.isEmpty()) {
-            return Optional.of(HttpStatus.FORBIDDEN);
+            return HttpStatus.FORBIDDEN;
         }
         HttpMethod methodToMatch = methodToMatchOptional.get();
 
-        Optional<? extends ArrayList<HttpMethod>> availableHttpMethods = (Optional<? extends ArrayList<HttpMethod>>) request.getAttribute(AVAILABLE_HTTP_METHODS, new ArrayList<HttpMethod>().getClass());
-        if (CorsUtil.isPreflightRequest(request) &&
-            availableHttpMethods.isPresent() &&
-            availableHttpMethods.get().stream().anyMatch(method -> method.equals(methodToMatch))) {
-            if (!hasAllowedHeaders(request, config)) {
-                return Optional.of(HttpStatus.FORBIDDEN);
+        if (hasAllowedHeaders(request, config)) {
+            List<UriRouteMatch<Object, Object>> anyUriRoutes = router.findAny(request);
+            if (anyUriRoutes.isEmpty()) {
```

Comment:
This is how it worked before, but I think that's not correct. `methodToMatchOptional` is for preflight check is OPTIONS and there is no such method in mocked tests

Reply:
Shouldt'n we fail OPTIONS call if there is no route present under that path?

Reply:
No, in that case the preflight should return a good response and the actual cors request should return the 404. This will allow the cors client to see that the route is missing. Otherwise it will just get a permission error